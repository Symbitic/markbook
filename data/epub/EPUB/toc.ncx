<?xml version="1.0" encoding="UTF-8"?>
<ncx xmlns="http://www.daisy.org/z3986/2005/ncx/" version="2005-1" xml:lang="en">
  <head>
    <meta name="dtb:depth" content="1"/>
    <meta name="dtb:totalPageCount" content="0"/>
    <meta name="dtb:maxPageNumber" content="0"/>
  </head>
  <docTitle><text>{{ title }}</text></docTitle>
  <docAuthor><text>{{ creator }}</text></docAuthor>
  <navMap>
    {{#each toc}}
    <navPoint id="navpoint-{{i}}" playOrder="{{i}}">
      <navLabel><text>{{title}}</text></navLabel>
      <content src="{{url}}" />
    </navPoint>
    {{/each}}
  </navMap>
</ncx>
