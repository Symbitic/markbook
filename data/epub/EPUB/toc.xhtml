<?xml version="1.0" encoding="UTF-8"?>
<html xmlns="http://www.w3.org/1999/xhtml" xmlns:epub="http://www.idpf.org/2007/ops">
  <head>
    <title>{{title}}</title>
    <meta charset="utf-8" />
  </head>
  <body>
    <nav xmlns:epub="http://www.idpf.org/2007/ops" epub:type="toc" id="toc">
      <ol>
        {{#each toc}}
        <li><a href="{{url}}">{{title}}</a></li>
        {{/each}}
      </ol>
    </nav>
  </body>
</html>
