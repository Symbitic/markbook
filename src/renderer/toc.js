export default function createToc (config) {
  const geturl = filename =>
    filename
      .replace(/README\.md$/, 'index.md')
      .replace(/\.md$/, '')
      .concat('.html')

  const chapter = (items, i) =>
    items.slice(0, i + 1).filter(item => item.level === 1).length

  const subchapter = (items, i) =>
    i -
    items.indexOf(
      items
        .slice(0, i + 1)
        .reverse()
        .find(i => i.level === 1)
    )

  const prefix = config.summary.prefix.map(item => ({
    title: item.title,
    url: geturl(item.url)
  }))

  const chapters = config.summary.chapters.map((item, i, items) => ({
    title: item.title,
    url: geturl(item.url),
    ch: `${chapter(items, i)}.${subchapter(items, i)}`.replace(/\.0$/, '.'),
    level: item.level
  }))

  const suffix = config.summary.suffix.map(item => ({
    title: item.title,
    url: geturl(item.url)
  }))

  return {
    prefix,
    chapters,
    suffix
  }
}
