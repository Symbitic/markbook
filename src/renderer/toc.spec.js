/* global expect */
import toc from './toc'

const basicPrefix = {
  prefix: [{ title: 'Markbook', url: 'README.md' }],
  chapters: [],
  suffix: []
}

const chapters = {
  prefix: [],
  chapters: [
    { title: 'Chapter 1', url: 'chapter-1.md', level: 1 },
    { title: 'Chapter 2', url: 'chapter-2.md', level: 1 },
    { title: 'Chapter 2.1', url: 'chapter-2-1.md', level: 2 },
    { title: 'Chapter 2.2', url: 'chapter-2-1.md', level: 2 },
    { title: 'Chapter 3', url: 'chapter-3.md', level: 1 }
  ],
  suffix: []
}

const case1 = ['basic prefix', basicPrefix]
const case2 = ['list of chapters', chapters]

const tests = [case1, case2]

describe.each(tests)('toc', (name, source) => {
  it(`should parse a ${name}`, () => {
    expect(toc({ summary: source })).toMatchSnapshot()
  })
})
