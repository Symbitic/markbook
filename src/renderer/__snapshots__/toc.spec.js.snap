// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toc should parse a basic prefix 1`] = `
Object {
  "chapters": Array [],
  "prefix": Array [
    Object {
      "title": "Markbook",
      "url": "index.html",
    },
  ],
  "suffix": Array [],
}
`;

exports[`toc should parse a list of chapters 1`] = `
Object {
  "chapters": Array [
    Object {
      "ch": "1.",
      "level": 1,
      "title": "Chapter 1",
      "url": "chapter-1.html",
    },
    Object {
      "ch": "2.",
      "level": 1,
      "title": "Chapter 2",
      "url": "chapter-2.html",
    },
    Object {
      "ch": "2.1",
      "level": 2,
      "title": "Chapter 2.1",
      "url": "chapter-2-1.html",
    },
    Object {
      "ch": "2.2",
      "level": 2,
      "title": "Chapter 2.2",
      "url": "chapter-2-1.html",
    },
    Object {
      "ch": "3.",
      "level": 1,
      "title": "Chapter 3",
      "url": "chapter-3.html",
    },
  ],
  "prefix": Array [],
  "suffix": Array [],
}
`;
